docs/modelplugin.md

```diff
@@ -0,0 +1,175 @@
+the configuration and state objects and attributes of a device. In practice a
+device's object model usually comprises of a number of YANG files including
+augments and deviations, and must be considered as a combined unit.
+
+In **onos-config** a set of these combined YANG files defining a particular
+version of a device type is known as a **model**.
+
+Over its lifecycle onos-config will have to deal with many different models as
+its scope is expanded and as devices go through new release cycles. To allow
+this models are loadable dynamically as plugins in the form of Linux or Mac
+shared object libraries **(\*.so)** using the YGOT library and are know as
+**Model Plugins**.
+
+The diagram shows the connection between the Model Plugin and the configuration
+store - linked by Device Type and Version. Effectively the primary key
+of the Model Registry is the Model Name and Version, whereas the primary key of
+the Configuration is the Device Name and Version. 
+
+![onos-config internals](images/onos-config-internals.png)
+
+## Structure of a Model Plugin
+A Model Plugin is mainly generated by the generator command from the YGOT
+project, and a wrapper **modelmain.go** implementing the ModelPlugin interface.
+They are compiled together with the **go build** command using the
+**-buildmode=plugin** option. 
+
+Many examples of Model Plugins are in the [../modelplugins](../modelplugins)
+folder of this project, and an example script
+[ModelGenerator.sh](../modelplugins/ModelGenerator.sh) is available for creating
+new plugins. 
+
+## YANG files
+The YANG files to be used with generator.go should be collected together in a
+folder and named in the style: **\<modulename>@\<latestrevision>.yang**
+
+Running the generator command in the form:
+```bash
+go run $GOPATH/src/github.com/openconfig/ygot/generator/generator.go \
+-path yang -output_file=$TYPEVERSION/$TYPEVERSIONPKG/generated.go -package_name=$TYPEVERSIONPKG \
+-generate_fakeroot $YANGLIST
+```
+will check all nested dependencies are present, **and** that the output is
+generated as a single file: **generated.go**.
+> Where $YANGLIST is a space separated list of YANG file names.
+> See ModelGenerator.sh for an example
+
+To visualize and further validate the collection of YANG files, the
+**[pyang](https://github.com/mbj4668/pyang)** tool can be used like:
+```bash
+pyang -f tree $YANGLIST
+```
+> Once the generator has run there is no need to persist the YANG files - the
+> generated.go file contains all the information in an object model.
+
+
+## modelmain.go definitions
+> Examples of these definitions are given in [ModelGenerator.sh](../modelplugins/ModelGenerator.sh)
+
+### modeltype
+This should be a name that defines the type of device, but should not include
+version. This name will be used later in the **Configuration** of the device.
+It should be between 4 and 40 chars and only include alphanumeric characters,
+'-', '_' and ':'.
+
+### modelversion
```

Comment:
for some reason from here nothing gets highlighted. I believe its due do the ` at line 70. can you doble check ?

Reply:
done